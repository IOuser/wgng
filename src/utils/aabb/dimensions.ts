export type Dimensions = {
    width: number;
    height: number;
};

export type Coordinates = {
    x: number;
    y: number;
};

export type FullDimensions = Dimensions & Coordinates;
