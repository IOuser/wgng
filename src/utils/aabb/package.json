{
    "private": true,
    "name": "aabb",
    "main": "aabb.ts",
    "typings": "aabb.ts"
}
