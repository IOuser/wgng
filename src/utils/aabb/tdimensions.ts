export type TDimensions = {
    width: number;
    height: number;
};

export type TFullDimensions = TDimensions & {
    x: number;
    y: number;
};
