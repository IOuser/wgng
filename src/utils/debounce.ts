export function debounce<F extends Function>(func: F, waitMilliseconds: number = 50): F {
    // tslint:disable-next-line:no-any
    let timeoutId: any;

    // tslint:disable-next-line:no-any
    return function(this: any, ...args: any[]): void {
        const doLater = () => {
            timeoutId = undefined;
            func.apply(this, args);
        };

        if (timeoutId !== undefined) {
            clearTimeout(timeoutId);
        }

        timeoutId = setTimeout(doLater, waitMilliseconds);

        if (timeoutId === undefined) {
            func.apply(this, args);
        }
    } as any; // tslint:disable-line:no-any
}
