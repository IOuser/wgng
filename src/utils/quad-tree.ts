export type XY = {
    x: number;
    y: number;
};

export const enum AABBPart {
    NW,
    NE,
    SW,
    SE,
}

export interface AABB {
    containsPoint(point: XY): boolean;
    intersectsAABB(other: AABB): boolean;
    subdivide(part: AABBPart): AABB;
    dims(): [number, number, number, number];
}

export class AxisAlignedBoundingBox implements AABB {
    private _center: XY;
    private _halfDimension: XY;

    public constructor(center: XY, halfDimension: XY) {
        this._center = center;
        this._halfDimension = halfDimension;
    }

    public containsPoint(point: XY): boolean {
        const { x, y } = point;

        const left = this._center.x - this._halfDimension.x;
        const right = this._center.x + this._halfDimension.x;
        const top = this._center.y - this._halfDimension.y;
        const bottom = this._center.y + this._halfDimension.y;

        return !(x < left || x > right || y < top || y > bottom);
    }

    public intersectsAABB(_aabb: AABB): boolean {
        return true;
    }

    public subdivide(part: AABBPart): AABB {
        const { x, y } = this._center;
        const { x: halfW, y: halfY } = this._halfDimension;
        const quadW = halfW / 2;
        const quadH = halfY / 2;
        const halfDimension = { x: quadW, y: quadH };

        switch (true) {
            case part === AABBPart.NW:
                return new AxisAlignedBoundingBox({ x: x - quadW, y: y - quadH }, halfDimension);
            case part === AABBPart.NE:
                return new AxisAlignedBoundingBox({ x: x + quadW, y: y - quadH }, halfDimension);
            case part === AABBPart.SW:
                return new AxisAlignedBoundingBox({ x: x - quadW, y: y + quadH }, halfDimension);
            case part === AABBPart.SE:
                return new AxisAlignedBoundingBox({ x: x + quadW, y: y + quadH }, halfDimension);
            default:
                throw Error('Part have incorrect value');
        }
    }

    public dims(): [number, number, number, number] {
        const { x, y } = this._center;
        const { x: halfW, y: halfY } = this._halfDimension;
        return [x - halfW, y - halfY, halfW * 2, halfY * 2];
    }
}

const enum Constants {
    QT_NODE_CAPACITY = 4,
}

export class QuadTree {
    private _boundary: AABB;
    private _points: XY[] = [];

    private _northWest: QuadTree | null = null;
    private _northEast: QuadTree | null = null;
    private _southWest: QuadTree | null = null;
    private _southEast: QuadTree | null = null;

    public constructor(boundary: AABB) {
        this._boundary = boundary;
    }

    public renderNodes(ctx: CanvasRenderingContext2D): void {
        if (this._northWest === null) {
            ctx.strokeRect(...this._boundary.dims());
            return;
        }

        this._northWest!.renderNodes(ctx);
        this._northEast!.renderNodes(ctx);
        this._southWest!.renderNodes(ctx);
        this._southEast!.renderNodes(ctx);
    }

    public reset(): void {
        this._points = [];
        this._northWest = null;
        this._northEast = null;
        this._southWest = null;
        this._southEast = null;
    }

    public queryRange(range: AABB): XY[] {
        // debugger
        const result: XY[] = [];

        if (!this._boundary.intersectsAABB(range)) {
            return result;
        }

        for (const point of this._points) {
            if (range.containsPoint(point)) {
                result.push(point);
            }
        }

        if (this._northWest === null) {
            return result;
        }

        return [
            ...result,
            ...this._northWest!.queryRange(range),
            ...this._northEast!.queryRange(range),
            ...this._southWest!.queryRange(range),
            ...this._southEast!.queryRange(range),
        ];
    }

    public insert(point: XY): boolean {
        if (!this._boundary.containsPoint(point)) {
            return false;
        }

        if (this._points.length < Constants.QT_NODE_CAPACITY) {
            this._points.push(point);
            return true;
        }

        if (this._northWest === null) {
            return this._subdivide(point);
        }

        // TODO: Add tests
        throw Error('waat');
        return false;
    }

    private _subdivide(newPoint: XY): boolean {
        this._northWest = new QuadTree(this._boundary.subdivide(AABBPart.NW));
        this._northEast = new QuadTree(this._boundary.subdivide(AABBPart.NE));
        this._southWest = new QuadTree(this._boundary.subdivide(AABBPart.SW));
        this._southEast = new QuadTree(this._boundary.subdivide(AABBPart.SE));

        for (const point of this._points) {
            if (
                this._northWest.insert(point) ||
                this._northEast.insert(point) ||
                this._southWest.insert(point) ||
                this._southEast.insert(point)
            ) {
                continue;
            }

            throw Error('Wrong sub tree range');
        }

        return (
            this._northWest.insert(newPoint) ||
            this._northEast.insert(newPoint) ||
            this._southWest.insert(newPoint) ||
            this._southEast.insert(newPoint)
        );
    }
}
