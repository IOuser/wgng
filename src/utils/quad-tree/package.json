{
    "private": true,
    "name": "quad-tree",
    "main": "quad-tree.ts",
    "typings": "quad-tree.ts"
}
