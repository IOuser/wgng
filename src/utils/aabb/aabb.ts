import { IPoint, Point } from 'utils/point';

import { TFullDimensions } from './tdimensions';

export const enum AABBPart {
    NW,
    NE,
    SW,
    SE,
}

// AxisAlignedBoundingBox
export interface IAABB {
    containsPoint(point: IPoint): boolean;
    intersectsAABB(other: IAABB): boolean;
    subdivide(part: AABBPart): IAABB;
    fullDimensions(): TFullDimensions;
}

export class AABB implements IAABB {
    private _center: IPoint;
    private _halfDimension: IPoint;

    public constructor(center: IPoint, halfDimension: IPoint) {
        this._center = center;
        this._halfDimension = halfDimension;
    }

    public containsPoint(point: IPoint): boolean {
        const { x, y } = point;

        const left = this._center.x - this._halfDimension.x;
        const right = this._center.x + this._halfDimension.x;
        const top = this._center.y - this._halfDimension.y;
        const bottom = this._center.y + this._halfDimension.y;

        return !(x < left || x > right || y < top || y > bottom);
    }

    public intersectsAABB(aabb: IAABB): boolean {
        const aDims = this.fullDimensions();
        const bDims = aabb.fullDimensions();

        const aLeft = aDims.x;
        const aRight = aDims.x + aDims.width;
        const aTop = aDims.y;
        const aBottom = aDims.y + aDims.height;

        const bLeft = bDims.x;
        const bRight = bDims.x + bDims.width;
        const bTop = bDims.y;
        const bBottom = bDims.y + bDims.height;

        if (aLeft > bRight || aRight < bLeft || aTop > bBottom || aBottom < bTop) {
            return false;
        }

        return true;
    }

    public subdivide(part: AABBPart): IAABB {
        const { x, y } = this._center;
        const { x: halfW, y: halfY } = this._halfDimension;
        const quadW = halfW / 2;
        const quadH = halfY / 2;
        const halfDimension = new Point(quadW, quadH);

        switch (true) {
            case part === AABBPart.NW:
                return new AABB(new Point(x - quadW, y - quadH), halfDimension);
            case part === AABBPart.NE:
                return new AABB(new Point(x + quadW, y - quadH), halfDimension);
            case part === AABBPart.SW:
                return new AABB(new Point(x - quadW, y + quadH), halfDimension);
            case part === AABBPart.SE:
                return new AABB(new Point(x + quadW, y + quadH), halfDimension);
            default:
                throw Error('Part have incorrect value');
        }
    }

    public fullDimensions(): TFullDimensions {
        const { x, y } = this._center;
        const { x: halfW, y: halfY } = this._halfDimension;
        return {
            x: x - halfW,
            y: y - halfY,
            width: halfW * 2,
            height: halfY * 2,
        };
    }
}
