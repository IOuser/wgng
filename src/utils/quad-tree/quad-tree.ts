import { IPoint } from 'utils/point';
import { IAABB, AABBPart } from 'utils/aabb';

const enum Constants {
    NodeCapacity = 4,
}

export class QuadTree {
    private _boundary: IAABB;
    private _points: IPoint[] = [];

    private _northWest: QuadTree | null = null;
    private _northEast: QuadTree | null = null;
    private _southWest: QuadTree | null = null;
    private _southEast: QuadTree | null = null;

    public constructor(boundary: IAABB) {
        this._boundary = boundary;
    }

    public renderNodes(ctx: CanvasRenderingContext2D): void {
        if (process.env.NODE_ENV !== 'development') {
            return;
        }

        if (this._northWest === null) {
            const { x, y, width, height } = this._boundary.fullDimensions();
            ctx.strokeRect(x, y, width, height);
            return;
        }

        this._northWest!.renderNodes(ctx);
        this._northEast!.renderNodes(ctx);
        this._southWest!.renderNodes(ctx);
        this._southEast!.renderNodes(ctx);
    }

    public reset(): void {
        this._points = [];
        this._northWest = null;
        this._northEast = null;
        this._southWest = null;
        this._southEast = null;
    }

    public queryRange(range: IAABB): IPoint[] {
        // debugger
        const result: IPoint[] = [];

        if (!this._boundary.intersectsAABB(range)) {
            return result;
        }

        for (const point of this._points) {
            if (range.containsPoint(point)) {
                result.push(point);
            }
        }

        if (this._northWest === null) {
            return result;
        }

        return [
            ...result,
            ...this._northWest!.queryRange(range),
            ...this._northEast!.queryRange(range),
            ...this._southWest!.queryRange(range),
            ...this._southEast!.queryRange(range),
        ];
    }

    public outerRange(range: IAABB): IPoint[] {
        const result: IPoint[] = [];

        // Add "isSubBound" method to AABB
        // if (this._boundary.intersectsAABB(range)) {
        //     return result;
        // }

        for (const point of this._points) {
            if (!range.containsPoint(point)) {
                result.push(point);
            }
        }

        if (this._northWest === null) {
            return result;
        }

        return [
            ...result,
            ...this._northWest!.outerRange(range),
            ...this._northEast!.outerRange(range),
            ...this._southWest!.outerRange(range),
            ...this._southEast!.outerRange(range),
        ];
    }

    // tslint:disable-next-line:cyclomatic-complexity
    public insert(point: IPoint): boolean {
        if (!this._boundary.containsPoint(point)) {
            return false;
        }

        if (this._points.length < Constants.NodeCapacity) {
            this._points.push(point);
            return true;
        }

        if (
            this._northWest === null ||
            this._northEast === null ||
            this._southWest === null ||
            this._southEast === null
        ) {
            this._northWest = new QuadTree(this._boundary.subdivide(AABBPart.NW));
            this._northEast = new QuadTree(this._boundary.subdivide(AABBPart.NE));
            this._southWest = new QuadTree(this._boundary.subdivide(AABBPart.SW));
            this._southEast = new QuadTree(this._boundary.subdivide(AABBPart.SE));
        }

        const points = this._points;
        while (points.length) {
            const currentPoint = points.shift();
            if (currentPoint === undefined) {
                break;
            }

            if (
                this._northWest.insert(currentPoint) ||
                this._northEast.insert(currentPoint) ||
                this._southWest.insert(currentPoint) ||
                this._southEast.insert(currentPoint)
            ) {
                continue;
            }

            throw Error('Wrong sub tree range');
        }

        return (
            this._northWest.insert(point) ||
            this._northEast.insert(point) ||
            this._southWest.insert(point) ||
            this._southEast.insert(point)
        );
    }
}
