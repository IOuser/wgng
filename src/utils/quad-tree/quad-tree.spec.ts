import { AABB } from 'utils/aabb';
import { Point } from 'utils/point';

import { QuadTree } from './quad-tree';

describe('QuadTree', () => {
    it('should match snapshot', () => {
        const qt = new QuadTree(new AABB(new Point(1, 1), new Point(1, 1)));

        qt.insert(new Point(0.25, 0.1));
        qt.insert(new Point(0.25, 0.2));
        qt.insert(new Point(0.25, 0.3));
        qt.insert(new Point(0.25, 0.4));
        qt.insert(new Point(0.25, 0.5));

        expect(qt).toMatchSnapshot();
    });
});
