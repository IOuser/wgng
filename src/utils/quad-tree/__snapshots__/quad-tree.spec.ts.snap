// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuadTree should match snapshot 1`] = `
QuadTree {
  "_boundary": AABB {
    "_center": Point {
      "x": 1,
      "y": 1,
    },
    "_halfDimension": Point {
      "x": 1,
      "y": 1,
    },
  },
  "_northEast": QuadTree {
    "_boundary": AABB {
      "_center": Point {
        "x": 1.5,
        "y": 0.5,
      },
      "_halfDimension": Point {
        "x": 0.5,
        "y": 0.5,
      },
    },
    "_northEast": null,
    "_northWest": null,
    "_points": Array [],
    "_southEast": null,
    "_southWest": null,
  },
  "_northWest": QuadTree {
    "_boundary": AABB {
      "_center": Point {
        "x": 0.5,
        "y": 0.5,
      },
      "_halfDimension": Point {
        "x": 0.5,
        "y": 0.5,
      },
    },
    "_northEast": QuadTree {
      "_boundary": AABB {
        "_center": Point {
          "x": 0.75,
          "y": 0.25,
        },
        "_halfDimension": Point {
          "x": 0.25,
          "y": 0.25,
        },
      },
      "_northEast": null,
      "_northWest": null,
      "_points": Array [],
      "_southEast": null,
      "_southWest": null,
    },
    "_northWest": QuadTree {
      "_boundary": AABB {
        "_center": Point {
          "x": 0.25,
          "y": 0.25,
        },
        "_halfDimension": Point {
          "x": 0.25,
          "y": 0.25,
        },
      },
      "_northEast": QuadTree {
        "_boundary": AABB {
          "_center": Point {
            "x": 0.375,
            "y": 0.125,
          },
          "_halfDimension": Point {
            "x": 0.125,
            "y": 0.125,
          },
        },
        "_northEast": null,
        "_northWest": null,
        "_points": Array [],
        "_southEast": null,
        "_southWest": null,
      },
      "_northWest": QuadTree {
        "_boundary": AABB {
          "_center": Point {
            "x": 0.125,
            "y": 0.125,
          },
          "_halfDimension": Point {
            "x": 0.125,
            "y": 0.125,
          },
        },
        "_northEast": null,
        "_northWest": null,
        "_points": Array [
          Point {
            "x": 0.25,
            "y": 0.1,
          },
          Point {
            "x": 0.25,
            "y": 0.2,
          },
        ],
        "_southEast": null,
        "_southWest": null,
      },
      "_points": Array [],
      "_southEast": QuadTree {
        "_boundary": AABB {
          "_center": Point {
            "x": 0.375,
            "y": 0.375,
          },
          "_halfDimension": Point {
            "x": 0.125,
            "y": 0.125,
          },
        },
        "_northEast": null,
        "_northWest": null,
        "_points": Array [],
        "_southEast": null,
        "_southWest": null,
      },
      "_southWest": QuadTree {
        "_boundary": AABB {
          "_center": Point {
            "x": 0.125,
            "y": 0.375,
          },
          "_halfDimension": Point {
            "x": 0.125,
            "y": 0.125,
          },
        },
        "_northEast": null,
        "_northWest": null,
        "_points": Array [
          Point {
            "x": 0.25,
            "y": 0.3,
          },
          Point {
            "x": 0.25,
            "y": 0.4,
          },
          Point {
            "x": 0.25,
            "y": 0.5,
          },
        ],
        "_southEast": null,
        "_southWest": null,
      },
    },
    "_points": Array [],
    "_southEast": QuadTree {
      "_boundary": AABB {
        "_center": Point {
          "x": 0.75,
          "y": 0.75,
        },
        "_halfDimension": Point {
          "x": 0.25,
          "y": 0.25,
        },
      },
      "_northEast": null,
      "_northWest": null,
      "_points": Array [],
      "_southEast": null,
      "_southWest": null,
    },
    "_southWest": QuadTree {
      "_boundary": AABB {
        "_center": Point {
          "x": 0.25,
          "y": 0.75,
        },
        "_halfDimension": Point {
          "x": 0.25,
          "y": 0.25,
        },
      },
      "_northEast": null,
      "_northWest": null,
      "_points": Array [],
      "_southEast": null,
      "_southWest": null,
    },
  },
  "_points": Array [],
  "_southEast": QuadTree {
    "_boundary": AABB {
      "_center": Point {
        "x": 1.5,
        "y": 1.5,
      },
      "_halfDimension": Point {
        "x": 0.5,
        "y": 0.5,
      },
    },
    "_northEast": null,
    "_northWest": null,
    "_points": Array [],
    "_southEast": null,
    "_southWest": null,
  },
  "_southWest": QuadTree {
    "_boundary": AABB {
      "_center": Point {
        "x": 0.5,
        "y": 1.5,
      },
      "_halfDimension": Point {
        "x": 0.5,
        "y": 0.5,
      },
    },
    "_northEast": null,
    "_northWest": null,
    "_points": Array [],
    "_southEast": null,
    "_southWest": null,
  },
}
`;
